# 01. Importing Libraries

In [1]:
# Importing libraries
import pandas as pd
import numpy as np
import os

# 02. Importing Data

In [2]:
# set file location as 'path' string
path = r'C:\Users\Logan\OneDrive\Desktop\Instacart Basket Analysis'
path

'C:\\Users\\Logan\\OneDrive\\Desktop\\Instacart Basket Analysis'

In [4]:
# import the 'products.csv' data set as df_prods
# import the 'orderswrangled.csv' data set as df_ords
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', '4.4_orders_wrangled.csv' ),index_col = False)

In [5]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [6]:
df_ords.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [40]:
df_ords.shape

(1048575, 7)

In [7]:
# remove added index column
df_ords = df_ords.drop(columns = ['Unnamed: 0'])

KeyError: "['Unnamed: 0'] not found in axis"

In [ ]:
df_ords.head()

In [ ]:
# Rename 'last_order_days' to 'days_since_prior_order
df_ords = df_ords.rename(columns = {'last_order_days' : 'days_since_prior_order'})

In [ ]:
df_ords.describe()

# 03. Mixed-Type Data (Data Consistency Checks)

In [ ]:
# Create a dataframe

df_test = pd.DataFrame()

In [ ]:
# Create a mixed-type column

df_test['mix'] = ['a', 'b', 1, True]

In [ ]:
df_test['standard'] = [4,5,6,12]

In [ ]:
df_test.head()

In [ ]:
# Check for mixed types

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

In [ ]:
# Fix mixed-type column
df_test['mix'] = df_test['mix'].astype('str')

In [ ]:
# Check for mixed types

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

# 04. Missing Values (Data Consistency Checks)

In [8]:
# Finding missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [9]:
# Create subset to view missing values
df_nan = df_prods.loc[df_prods['product_name'].isnull() == True]

In [10]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [11]:
df_prods.shape

(49693, 5)

In [12]:
# Create a new data frame called df_prods_clean, filtering out missing values
df_prods_clean = df_prods.loc[df_prods['product_name'].isnull() == False]

In [13]:
# Check that the number of rows has decreased
df_prods_clean.shape

(49677, 5)

In [14]:
# Check that there are no missing values in newly created data frame
df_prods_clean.isnull().sum()

product_id       0
product_name     0
aisle_id         0
department_id    0
prices           0
dtype: int64

# 05. Duplicates (Data Consistency Checks)

In [15]:
# Create new subset of df_prods_clean to look for duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [16]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [17]:
df_prods_clean.shape

(49677, 5)

In [18]:
# Create new dataframe that does not include the duplicates identified
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [19]:
df_prods_clean_no_dups.shape

(49672, 5)

In [20]:
# Check that no duplicates exist
df_prods_clean_no_dups[df_prods_clean_no_dups.duplicated()]

Empty DataFrame
Columns: [product_id, product_name, aisle_id, department_id, prices]
Index: []

In [21]:
# Export dataframe into prepared data folder
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

# 06. 4.5 Task Procedures

In [22]:
# Run the df.describe() function on your df_prods dataframe. 
# Using your knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

1. prices column in df_prods dataframe should be investigated

1a. max price of 99999.00, however this is an unrealistic price for instacart data. This is clearly an outlier 75% quartile is 11.20. This could be disproportionately affecting the mean price.

In [23]:
# Check for mixed-type data in your df_ords dataframe
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [24]:
df_ords.dtypes

order_id                    int64
user_id                     int64
order_number                int64
order_day_of_week           int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

In [25]:
# Run a check for missing values in your df_ords dataframe
df_ords.isnull().sum()

order_id                      0
user_id                       0
order_number                  0
order_day_of_week             0
order_hour_of_day             0
days_since_prior_order    63100
dtype: int64

In [26]:
df_ords.shape

(1048575, 6)

In a markdown cell, report your findings and propose an explanation for any missing values you find.

206209 missing values in days_since_prior_order. While it would be best to check with client on any missing data, this is most likely a result of customers purchasing their first order and as a result this column would hold 0 or missing values.

In [27]:
# Address the missing values using an appropriate method
df_ords_new = df_ords
# Add indicator column to denote True if days_since_prior_order is NaN as first_order
df_ords_new['first_order'] = df_ords['days_since_prior_order'].isnull() == True

In [28]:
df_ords_new[df_ords_new['days_since_prior_order'].isnull() == True]

order_id  user_id  order_number  order_day_of_week  \
0         2539329        1             1                  2   
11        2168274        2             1                  2   
26        1374495        3             1                  1   
39        3343014        4             1                  6   
45        2717275        5             1                  3   
...           ...      ...           ...                ...   
1048523   1591766    63096             1                  0   
1048530   3391694    63097             1                  2   
1048541   1270497    63098             1                  3   
1048556   1877750    63099             1                  0   
1048570   1805240    63100             1                  0   

         order_hour_of_day  days_since_prior_order  first_order  
0                        8                     NaN         True  
11                      11                     NaN         True  
26                      14                     NaN         True  
39                      11                     NaN         True  
45                      12                     NaN         True  
...                    ...                     ...          ...  
1048523                 20                     NaN         True  
1048530                 14                     NaN         True  
1048541                 16                     NaN         True  
1048556                 10                     NaN         True  
1048570                 11                     NaN         True  

[63100 rows x 7 columns]

In [29]:
df_ords_new.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  first_order  
0                     NaN         True  
1                    15.0        False  
2                    21.0        False  
3                    29.0        False  
4                    28.0        False

In a markdown cell, explain why you used your method of choice.

Created a new variable 'first_order' that actas like a flag based on the missing value 'days_since_prior_order' to denote if it was the first order for a customer. These missing values have importance in showing first time customers and should not be removed.

In [30]:
# Run a check for duplicate values in your df_ords data.
df_ords_dups = df_ords_new[df_ords_new.duplicated()]

In [31]:
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, order_day_of_week, order_hour_of_day, days_since_prior_order, first_order]
Index: []

In a markdown cell, report your findings and propose an explanation for any duplicate values you find.

No duplicates found in df_ords_new dataframe.

In [32]:
df_ords_new.shape

(1048575, 7)

In [33]:
df_ords_new_no_dups = df_ords_new.drop_duplicates()

In [34]:
df_ords_new_no_dups.shape

(1048575, 7)

In [35]:
df_ords_new_no_dups.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  first_order  
0                     NaN         True  
1                    15.0        False  
2                    21.0        False  
3                    29.0        False  
4                    28.0        False

In [36]:
df_ords_new_no_dups.shape

(1048575, 7)

In [38]:
df_ords_updated = df_ords_new_no_dups.drop(columns = ['first_order'])

In [39]:
df_ords_updated.to_csv(os.path.join(path, '02 Data', 'Prepared Data', '4.5_orders_updated.csv'))

In [ ]:
# Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.
df_ords_new_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))